### PR TITLE
CI: Move sql builds from Travis to Github actions

### PR DESCRIPTION
**Pull request by fangchenli**
try to move builds with databases to GitHub actions

### CODE REVIEW

File: .github/workflows/database.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      run: 
+        sudo systemctl enable mysql.service
+        sudo systemctl start mysql.service
+
+    - name: Setup environment and build pandas
+      run: ci/setup_env.sh
+      env:
+        POSTGRES_HOST: localhost
+        POSTGRES_PORT: ${{ job.services.postgres.ports[5432] }} 
+      if: always()
```
Comment (MarcoGorelli) [inline .github/workflows/database.yml:61]: if we're going to use githubactions, we may get better environment setup time using [conda incubator](https://github.com/conda-incubator/setup-miniconda)

e.g.

```
      - uses: actions/checkout@v2
      - name: Cache conda
        uses: actions/cache@v1
        env:
          # Increase this value to reset cache if environment-dev-py36.yml has not changed
          CACHE_NUMBER: 0
        with:
          path: ~/conda_pkgs_dir
          key: ${{ runner.os }}-conda-${{ env.CACHE_NUMBER }}-${{
            hashFiles('environment.yml') }}
      - uses: conda-incubator/setup-miniconda@v2
        with:
          activate-environment: pandas-dev
          channel-priority: strict
          environment-file: environment.yml
          use-only-tar-bz2: true # IMPORTANT: This needs to be set for caching to work properly!
```

**Comment by simonjayhawkins** [comment]
marking this as 1.2 as any changes to the CI will also need to be backported. but not a blocker for 1.2.

File: .github/workflows/database.yml [hunk 2]
```diff
@@ -0,0 +1,94 @@
+name: Database
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches:
+      - master
+      - 1.1.x
```
Comment (simonjayhawkins) [inline .github/workflows/database.yml:9]: 1.2.x is now the active backport branch

**Comment by jreback** [comment]
looks good, @fangchenli can you make the change indicated by @simonjayhawkins merge master and ping on green (possible numpy-dev is still failing but that's ok)

**Comment by fangchenli** [comment]
I suspect SQL related tests did not run in this build. Let me double check.

**Comment by fangchenli** [comment]
The py38-slow build is not running any SQL tests. Doesn't it suppose to? If not, we could move it to Azure.

**Comment by jreback** [comment]
> The py38-slow build is not running any SQL tests. Doesn't it suppose to? If not, we could move it to Azure.

umm, whereever we have sqlalchemy installed should run (if it ins't we should add)

**Comment by jreback** [comment]
i think the npdev build is fixed on master if you can merge master

**Comment by fangchenli** [comment]
> > The py38-slow build is not running any SQL tests. Doesn't it suppose to? If not, we could move it to Azure.
> 
> umm, whereever we have sqlalchemy installed should run (if it ins't we should add)

Sqlalchemy is installed. I've commented out all SQL setup code and removed sqlalchemy. And the number of tests run is the same as before. So I think none of the SQL tests were run on this build.

**Comment by jreback** [comment]
interesting

ahh this a slow build - thought they were only slow tests

ok as long as we have sufficient coverage on sal tests its fine (and can revisit that later anyhow)

if u would open an issue

**Comment by jreback** [comment]
@fangchenli CI is currently failing on master (for other reasons). but ideally like to get this in. can you summarize what this is accomplishing:

- moving all database builds to azure
- enabling sql tests on 3.8_slow?
- removing builds from travis? 
what is left after this PR

**Comment by fangchenli** [comment]
@jreback The goal of this PR is to move all database builds to GitHub actions. After this, what left in Travis is the 3.8_slow #38429 and arm64.

**Comment by fangchenli** [comment]
And I'll have time to work on this after tomorrow.

**Comment by jreback** [comment]
same on this one. will be putting in 1.2.x

**Comment by jreback** [comment]
ping when ready with this one

**Comment by jreback** [comment]
and if you can merge master on this one

**Comment by jreback** [comment]
@fangchenli status of this one?

**Comment by fangchenli** [comment]
> @fangchenli status of this one?

Still working on it. It's harder than I thought.

**Comment by simonjayhawkins** [comment]
> Still working on it. It's harder than I thought.

moving to 1.2.1

**Comment by fangchenli** [comment]
The SQL databases are working correctly. Two issues remain.

1. We are currently testing the minimum support version of several packages in the locale build.

    - blosc=1.15.0
    - fastparquet=0.3.2
    - xarray=0.12.3
    - psycopg2=2.7
    - pymysql=0.7.11
    - sqlalchemy=1.3.0

They are causing some conflicts, and conda can not resolve them. It's failing on Travis too. I unpinned all of them for now.

2.  There are lots of failed tests and `TypeError: '<' not supported between instances of 'str' and 'int'` (some are from pyarrow and xarray). I'm not sure what is causing this many errors. They are not sql-related.

**Comment by jreback** [comment]
ok see if u can find the min pins to move these pins

likely the errors are related to other deps moving

**Comment by jreback** [comment]
you can try looking back in the build logs on master and see when this job first started failing - compare the before env and maybe can see what changes (and what we need to pin)

alt can try to create the env localky - conda has some verbose options to let u know what may is in conflict

**Comment by fangchenli** [comment]
> you can try looking back in the build logs on master and see when this job first started failing - compare the before env and maybe can see what changes (and what we need to pin)
> 
> alt can try to create the env localky - conda has some verbose options to let u know what may is in conflict

The conflict is caused by blosc. Increasing its minimum version to 1.17.0 fixed it.

**Comment by fangchenli** [comment]
Some of the errors look like #30333.

**Comment by jreback** [comment]
was the numpy version bumped?

also ok to omit blosc entirely

**Comment by fangchenli** [comment]
> was the numpy version bumped?
> 
> also ok to omit blosc entirely

numpy version was not specified in this build.

**Comment by fangchenli** [comment]
Using checkout v1 fixed most of those errors and test failures. But some remain

```bash
=========================== short test summary info ============================
FAILED pandas/tests/io/test_user_agent.py::test_server_and_default_headers[ParquetFastParquetUserAgentResponder-read_parquet-34273-fastparquet]
FAILED pandas/tests/io/test_user_agent.py::test_server_and_custom_headers[ParquetFastParquetUserAgentResponder-read_parquet-34275-fastparquet]
FAILED pandas/tests/io/test_user_agent.py::test_to_parquet_to_disk_with_storage_options[fastparquet]
FAILED pandas/tests/util/test_show_versions.py::test_show_versions - Attribut...
ERROR pandas/tests/io/test_fsspec.py - AttributeError: module 'numba' has no ...
ERROR pandas/tests/io/test_fsspec.py - AttributeError: module 'numba' has no ...
ERROR pandas/tests/io/test_gcs.py - AttributeError: module 'numba' has no att...
ERROR pandas/tests/io/test_gcs.py - AttributeError: module 'numba' has no att...
ERROR pandas/tests/io/test_parquet.py - AttributeError: module 'numba' has no...
ERROR pandas/tests/io/test_parquet.py - AttributeError: module 'numba' has no...
= 4 failed, 146993 passed, 3499 skipped, 1084 xfailed, 63 xpassed, 7 warnings, 6 errors in 1331.13s (0:22:11) =
```

**Comment by jreback** [comment]
can u merge master

**Comment by fangchenli** [comment]
All those errors are related to https://github.com/dask/fastparquet/issues/539. We need to either bump min fastparquet version to 0.4.0 or pin numba version to <0.52

**Comment by jreback** [comment]
> All those errors are related to [dask/fastparquet#539](https://github.com/dask/fastparquet/issues/539). We need to either bump min fastparquet version to 0.4.0 or pin numba version to <0.52

let's bump fastparquet & add a note in the whatsnew (do in 1.2.1 is fine)

File: .travis.yml [hunk 3]
```diff
@@ -16,50 +16,24 @@ services:
...
-    - env:
-        # Enabling Deprecations when running tests
-        # PANDAS_TESTING_MODE="deprecate" causes DeprecationWarning messages to be displayed in the logs
-        # See pandas/_testing.py for more details.
-        - JOB="3.7, coverage" ENV_FILE="ci/deps/travis-37-cov.yaml" PATTERN="((not slow and not network and not clipboard) or (single and db))" PANDAS_TESTING_MODE="deprecate" COVERAGE=true SQL="1"
-      services:
-        - mysql
-        - postgresql
-
-  allow_failures:
```
Comment (jreback) [inline .travis.yml:57]: still make this allow_failures so the build turns green :-<

File: doc/source/getting_started/install.rst [hunk 4]
```diff
@@ -263,10 +263,10 @@ Jinja2                    2.10               Conditional formatting with DataFra
...
 PyQt5                                        Clipboard I/O
 PyTables                  3.5.1              HDF5-based reading / writing
-SQLAlchemy                1.2.8              SQL support for databases other than sqlite
+SQLAlchemy                1.3.0              SQL support for databases other than sqlite
 SciPy                     1.12.0             Miscellaneous statistical functions
 xlsxwriter                1.0.2              Excel writing
-blosc                     1.15.0             Compression for HDF5
+blosc                     1.17.0             Compression for HDF5
 fsspec                    0.7.4              Handling files aside from local and HTTP
 fastparquet               0.3.2              Parquet reading / writing
```
Comment (jreback) [inline doc/source/getting_started/install.rst:271]: yep bump fastparquet is ok

File: pandas/tests/io/test_sql.py [hunk 5]
```diff
@@ -1334,7 +1334,7 @@ def test_sql_open_close(self):
 
     @pytest.mark.skipif(SQLALCHEMY_INSTALLED, reason="SQLAlchemy is installed")
     def test_con_string_import_error(self):
-        conn = "mysql://root@localhost/pandas_nosetest"
+        conn = "mysql://root@localhost:3306/pandas"
```
Comment (jreback) [inline pandas/tests/io/test_sql.py:1337]: where is this magic number from? can you make it a class level / constant

**Comment by jreback** [comment]
thanks @fangchenli really monumental here. !

a couple of followon things if you can.

- add the build versions printing for these 2 builds
- for the -cov build. i know we are attempting to get coverage here, but anything we can remove / reduce here? e.g. it would make sense to split this into 2 builds one for slow and one for the fast. though we *already* have these covered (except for the slow / db things). so maybe can just call this a slow build and repurpose one of the azure builds (38 linux maybe) as a coverage build?

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Comment by jbrockmendel** [comment]
the 1.3.0 whatsnew still lists 0.3.2 as the minimum version for fasparquet.  should this be updated?

**Comment by fangchenli** [comment]
> the 1.3.0 whatsnew still lists 0.3.2 as the minimum version for fasparquet. should this be updated?

just opened a new PR for that #40424
